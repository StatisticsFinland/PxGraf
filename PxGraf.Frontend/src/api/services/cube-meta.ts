/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { ICubeMeta } from "types/cubeMeta";
import { defaultQueryOptions } from "utils/ApiHelpers";

/**
 * Interface for cube metadata result. Cube is a multi-dimensional result of Px file data filtered by a query.
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {ICubeMeta} data - The cube meta data as an @see {@link ICubeMeta} object.
 */
export interface ICubeMetaResult {
    isLoading: boolean;
    isError: boolean;
    data: ICubeMeta;
}

const fetchCubeMeta = async(idStack: string[]): Promise<ICubeMeta> => {
    const client = new ApiClient();
    const url = `creation/cube-meta/${idStack.join('/')}`;

    return await client.getAsync(url);
}

export const useCubeMetaQuery = (idStack: string[]): ICubeMetaResult => {
    return useQuery(
        ['cube-meta', idStack],
        () => fetchCubeMeta(idStack),
        defaultQueryOptions
    );
}