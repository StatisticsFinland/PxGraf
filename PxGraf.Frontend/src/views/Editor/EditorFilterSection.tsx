import VariableSelectionList from 'components/VariableSelection/VariableSelectionList';
import React from 'react';
import { Box } from '@mui/material';
import { EditorContext } from 'contexts/editorContext';
import styled from 'styled-components';
import { IVariable } from 'types/cubeMeta';
import { Query } from 'types/query';

interface EditorFilterSectionProps {
    variables: IVariable[],
    resolvedVariableCodes: { [key: string]: string[] }
    queries: Query
    width?: number
    maxWidthPercentage?: number
}

const SelectorWrapper = styled(Box)<{width: number, $maxWidthPercentage: number}>`
  max-width: ${props => props.$maxWidthPercentage}%;
  flex: 0 1 ${props => props.width}px;
  position: fixed;
  top: 0;
  width: ${props => props.width}px;
  height: calc(100vh - 100px);
  margin-top: 98px;
  overflow: auto;
  padding-bottom: 50px;
  background-color: white;
  z-index: 5;
  border-top: thin solid rgba(0, 0, 0, 0.12);

  @media (max-width: 980px) {
    margin-top: 144px;
  }
`;

/**
 * Component for the filter section in the editor. Contains @see {@link VariableSelectionList} for each variable for filtering values and defining selectable variables.
 * @param {IVariable[]} variables Variables available for the table.
 * @param {{[key:string]: string[]}} resolvedVariableCodes Codes for the resolved variable values.
 * @param {Query} queries Object that contains variable queries
 * @param {number} width Width of the variable filter section
 * @param {number} maxWidthPercentage Maximum width of the variable filter section on the whole window defined in percentages
 */
export const EditorFilterSection: React.FC<EditorFilterSectionProps> = ({ variables, resolvedVariableCodes, queries, width, maxWidthPercentage }) => {
    
    const { setQuery } = React.useContext(EditorContext);

    return (
        <SelectorWrapper width={width} $maxWidthPercentage={maxWidthPercentage}>
            <VariableSelectionList
                variables={variables}
                resolvedVariableCodes={resolvedVariableCodes}
                query={queries}
                onQueryChanged={setQuery}
            />
        </SelectorWrapper>
    )
}

export default EditorFilterSection;