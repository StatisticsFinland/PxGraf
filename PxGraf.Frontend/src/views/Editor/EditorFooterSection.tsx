import React from 'react';
import { Box, Button } from '@mui/material';
import SaveIcon from '@mui/icons-material/Save';
import { EditorContext } from 'contexts/editorContext';
import { useTranslation } from 'react-i18next';
import styled from 'styled-components';
import InfoBubble from 'components/InfoBubble/InfoBubble';

const FooterBtnWrapper = styled(Box)`
    grid-area: 'footer';
    display: flex;
    justify-content: flex-end;
    align-items: center;
    padding: 16px;
`;


export const EditorFooterSection: React.FC = () => {
    
    const { setSaveDialogOpen } = React.useContext(EditorContext);
    
    const { t } = useTranslation();
    
    return(
        <FooterBtnWrapper>
            <InfoBubble info={t('infoText.save')} ariaLabel={t("editor.save")} />
            <Button variant="contained" startIcon={<SaveIcon />} onClick={() => setSaveDialogOpen(true)}>
                {t("editor.save")}
            </Button>
        </FooterBtnWrapper>
    );
}

export default EditorFooterSection;