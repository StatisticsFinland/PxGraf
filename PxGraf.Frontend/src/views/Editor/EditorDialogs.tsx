import React from 'react';

import { SaveDialog } from 'components/SaveDialog/SaveDialog';
import { SaveResultDialog } from 'components/SaveResultDialog/SaveResultDialog';
import { EditorContext } from 'contexts/editorContext';
import { ISaveQueryResult } from 'api/services/queries';

interface IEditorDialogsProps {
    saveQueryMutation: ISaveQueryResult;
}

export const EditorDialogs: React.FC<IEditorDialogsProps> = ({saveQueryMutation}) => {

    /* istanbul ignore next */
    const saveQueryAndShowResult = (archive: boolean) => {
        saveQueryMutation.mutate(archive);
        setSaveDialogOpen(false);
        setSaveResultDialogOpen(true);
    }

    const {saveDialogOpen, setSaveDialogOpen} = React.useContext(EditorContext);
    const [saveResultDialogOpen, setSaveResultDialogOpen] = React.useState(false);
    
    return (
        <>
            <SaveDialog open={saveDialogOpen} onClose={() => setSaveDialogOpen(false)} onSave = {saveQueryAndShowResult}/>
            <SaveResultDialog
            open={saveResultDialogOpen}
            onClose={() => setSaveResultDialogOpen(false)}
            mutation = {saveQueryMutation}
            />
        </>
    );
}

export default EditorDialogs;