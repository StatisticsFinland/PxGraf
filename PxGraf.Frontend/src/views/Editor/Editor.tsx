import React, { useEffect } from 'react';
import { useParams, useLocation } from "react-router-dom";
import { useTranslation } from 'react-i18next';

import { Box, Stack, Divider, Container, CircularProgress, Alert } from '@mui/material';

import { EditorContext } from 'contexts/editorContext';
import { getDefaultQueries, resolveVariables } from 'utils/editorHelpers';
import { getValidatedSettings } from 'utils/ChartSettingHelpers';
import EditorFilterSection from './EditorFilterSection';
import EditorFooterSection from './EditorFooterSection';
import EditorPreviewSection from './EditorPreviewSection';
import EditorMetaSection from './EditorMetaSection';
import EditorDialogs from './EditorDialogs';
import styled from 'styled-components';
import { useCubeMetaQuery } from 'api/services/cube-meta';
import { useDefaultHeaderQuery } from 'api/services/default-header';
import { useResolveVariableFiltersQuery } from 'api/services/filter-dimension';
import { useVisualizationOptionsQuery } from 'api/services/visualization-rules';
import { IFetchSavedQueryResponse, useSaveMutation } from 'api/services/queries';
import { VisualizationType } from 'types/visualizationType';
import { useQueryInfoQuery } from 'api/services/query-info';
import { extractCubeQuery, extractQuery } from 'utils/ApiHelpers';
import { useNavigationContext } from 'contexts/navigationContext';
import { useValidateTableMetadataQuery, IValidateTableMetaDataResult } from 'api/services/validate-table-metadata';
import { UiLanguageContext } from 'contexts/uiLanguageContext';
import { IDimension } from '../../types/cubeMeta';

//Used to set the width of the variable selection and preview margin in pixels
const variableSelectionWidth = 450;
//Max width in % for variable selection and preview margin
const variableSelectionMaxWidthPercentage = 33;

const MetaPreviewSectionWrapper = styled(Box)`
    flex: 2 0;
    display: grid;
    gap: 0;
    grid-template-columns: 1fr;
    grid-template-rows: auto auto 1fr auto auto;
    grid-template-areas: 'parameters' 'parameters-preview-div' 'preview' 'preview-footer-div' 'footer';
    min-height: 100%;
    margin-left: ${variableSelectionWidth}px;

    @media (max-width: ${variableSelectionWidth / (variableSelectionMaxWidthPercentage / 100) }px) {
        margin-left: ${variableSelectionMaxWidthPercentage}%;
    }
`;

const PreviewDivider = styled(Divider)`
  grid-area: 'parameters-preview-div';
`;

const FooterDivider = styled(Divider)`
  grid-area: 'preview-footer-div';
`;

const CubeMetaAlert = styled(Alert)`
  margin: 32px;
`;

/**
 * Main component for the editor view.
 * Contains the @see {@link EditorFilterSection} for defining values for each variable, @see {@link EditorMetaSection} for editing the displayed meta data, @see {@link EditorPreviewSection} for previewing the chart and @see {@link EditorFooterSection} for saving a visualization.
 */
export const Editor = () => {

    const location = useLocation();
    const { result }: { result: IFetchSavedQueryResponse } = location?.state ? location.state : { result: null };

    React.useEffect(() => {
        document.title = `${t("pages.editor")} | PxGraf`;
    }, []);

    // statemanagement
    const {
        cubeQuery,
        query,
        selectedVisualizationUserInput,
        visualizationSettingsUserInput,
        defaultSelectables,
        setQuery,
        setCubeQuery,
        setVisualizationSettingsUserInput,
        setSelectedVisualizationUserInput,
        setDefaultSelectables
    } = React.useContext(EditorContext);

    const { setTablePath } = useNavigationContext();

    useEffect(() => {
        if (result) {
            setQuery(extractQuery(result));
            setCubeQuery(extractCubeQuery(result));
            setVisualizationSettingsUserInput(result.settings);
            setSelectedVisualizationUserInput(result.settings.selectedVisualization);
            setDefaultSelectables(result.settings.defaultSelectableVariableCodes);
        }
    }, [result]);

    // hooks and support functions
    const { t } = useTranslation();
    const params = useParams();
    const pathStr = params["*"];
    const path = React.useMemo(
        () => pathStr.split("/").filter(p => p.length > 0),
        [pathStr]
    );

    useEffect(() => {
        if (path.length) {
            setTablePath(path);
        }
    }, [path]);

    // queries and other functionality
    const tableValidityResponse = useValidateTableMetadataQuery(path);
    const isTableInvalid = (response: IValidateTableMetaDataResult) => {
        if (response.data) {
            return !response.data.allVariablesContainValues || !response.data.tableHasContentVariable || !response.data.tableHasTimeVariable
        }
        else {
            return !response.isLoading;
        }
    }

    const tableIsInvalid = isTableInvalid(tableValidityResponse);
    const cubeMetaResponse = useCubeMetaQuery(path);

    const { language, languageTab, setLanguageTab, uiContentLanguage, setUiContentLanguage } = React.useContext(UiLanguageContext);

    const contentLanguages: string[] = cubeMetaResponse.data ? cubeMetaResponse.data.AvailableLanguages : [];

    useEffect(() => {
        if (!contentLanguages || contentLanguages.length == 0) {
            return;
        }
        if (contentLanguages.includes(language)) {
            setUiContentLanguage(language);
        }
        else if (!uiContentLanguage) {
            const contentLanguage = contentLanguages.includes(languageTab) ? languageTab : contentLanguages[0];
            setLanguageTab(contentLanguages[0]);
            setUiContentLanguage(contentLanguage)
        }
    }, [language, contentLanguages]);

    const dimensions: IDimension[] = cubeMetaResponse.data?.Dimensions ?? [];

    const modifiedQuery = React.useMemo(() => {
        if (query != null) {
            return query;
        }
        else if (dimensions != null) {
            return getDefaultQueries(dimensions);
        }
        else {
            return null;
        }
    }, [query, cubeMetaResponse.data]);

    const defaultHeaderResponse = useDefaultHeaderQuery(path, modifiedQuery);
    const queryInfoResponse = useQueryInfoQuery(path, modifiedQuery, cubeQuery);

    const resolvedVariableCodesResponse = useResolveVariableFiltersQuery(path, modifiedQuery);
    const resolvedVariableCodes = React.useMemo(() => {
        if (resolvedVariableCodesResponse.data != null) {
            return resolvedVariableCodesResponse.data;
        }
        else if (dimensions != null) {
            const varCodesNoVals = {}
            dimensions.forEach(v => {
                varCodesNoVals[v.Code] = [];
            })
            return varCodesNoVals;
        }
        else {
            return {};
        }
    }, [resolvedVariableCodesResponse, cubeMetaResponse.data]);
    const resolvedVariables = React.useMemo(() => {
        if (cubeMetaResponse.data != null && dimensions != null) {
            return resolveVariables(dimensions, resolvedVariableCodes);
        }
        else {
            return null;
        }
    }, [cubeMetaResponse.data, resolvedVariableCodes]);
    const selectedVisualization = React.useMemo(() => {
        if (queryInfoResponse.data?.validVisualizations.length > 0) {
            if (queryInfoResponse.data?.validVisualizations.includes(selectedVisualizationUserInput)) {
                return selectedVisualizationUserInput;
            }
            else {
                return queryInfoResponse.data?.validVisualizations[0] as VisualizationType;
            }
        }
        return null;
    }, [queryInfoResponse.data?.validVisualizations, selectedVisualizationUserInput]);
    const visualizationRulesResponse = useVisualizationOptionsQuery(
        path,
        modifiedQuery,
        selectedVisualization,
        // We should NOT use visualizationSettingsUserInput but since getDefaultSettings always return pivotRequested as false (or null) we can getaway with this cheat.
        visualizationSettingsUserInput?.pivotRequested ?? false,
    );
    const visualizationSettings = React.useMemo(() => {
        if (selectedVisualization != null && visualizationRulesResponse.data?.sortingOptions != null && resolvedVariables != null) {
            const result = getValidatedSettings(visualizationSettingsUserInput, selectedVisualization, visualizationRulesResponse.data.sortingOptions, resolvedVariables, modifiedQuery);
            if (defaultSelectables && Object.keys(defaultSelectables).length > 0) {
                result.defaultSelectableVariableCodes = defaultSelectables;
            } else {
                result.defaultSelectableVariableCodes = null;
            }
            return result;
        }
        else {
            return null;
        }
    }, [visualizationSettingsUserInput, selectedVisualization, visualizationRulesResponse.data?.sortingOptions, resolvedVariables, modifiedQuery, defaultSelectables]);
    const saveQueryMutation = useSaveMutation(path, modifiedQuery, cubeQuery, selectedVisualization, visualizationSettings);

    const errorContainer = (errorMessage: string) => {
        return (
            <Container sx={{ display: 'flex', alignItems: 'flex-start', justifyContent: 'center' }}>
                <CubeMetaAlert id="mainContent" severity="error">{errorMessage}</CubeMetaAlert>
            </Container>
        );
    }

    // return spinner or alert depending if results are loading or returns error 
    if (cubeMetaResponse.isLoading || tableValidityResponse.isLoading) {
        return (
            <Container sx={{ display: 'flex', alignItems: 'center', justifyContent: 'center' }}>
                <CircularProgress id="mainContent" />
            </Container>
        );
    }
    else if (tableIsInvalid || cubeMetaResponse.isError || !cubeMetaResponse?.data || !dimensions) {
        const errorWithCubeMeta = cubeMetaResponse.isError || !cubeMetaResponse?.data || !dimensions;
        const errorConditionsAndMessages = [
            { condition: tableValidityResponse.isError || (errorWithCubeMeta && tableValidityResponse.data?.allVariablesContainValues), message: t("error.contentLoad") },
            { condition: !tableValidityResponse.data?.tableHasContentVariable, message: t("error.contentVariableMissing") },
            { condition: !tableValidityResponse.data?.tableHasTimeVariable, message: t("error.timeVariableMissing") },
            { condition: !tableValidityResponse.data?.allVariablesContainValues, message: t("error.variablesMissingValues") }
        ];
        const errorMessages = errorConditionsAndMessages
            .filter(item => item.condition)
            .map(item => item.message);

        const errorMessage = errorMessages.join(" ");
        return errorContainer(errorMessage);
    }

    // return the actual component
    return (
        <Stack direction="row">
            <EditorFilterSection
                variables={dimensions}
                resolvedVariableCodes={resolvedVariableCodes}
                queries={modifiedQuery}
                width={variableSelectionWidth}
                maxWidthPercentage={variableSelectionMaxWidthPercentage}
            />
            <Divider orientation="vertical" />
            <MetaPreviewSectionWrapper>
                <EditorMetaSection
                    defaultHeaderResponse={defaultHeaderResponse}
                    visualizationRulesResponse={visualizationRulesResponse}
                    queryInfo={queryInfoResponse.data}
                    resolvedVariables={resolvedVariables}
                    selectedVisualization={selectedVisualization}
                    settings={visualizationSettings}
                    variableQuery={modifiedQuery}
                    contentLanguages={contentLanguages}
                />
                <PreviewDivider />
                <EditorPreviewSection
                    path={path}
                    query={modifiedQuery}
                    queryInfo={queryInfoResponse.data}
                    selectedVisualization={selectedVisualization}
                    visualizationSettings={visualizationSettings}
                />
                <FooterDivider />
                <EditorFooterSection />
                <EditorDialogs saveQueryMutation={saveQueryMutation} />
            </MetaPreviewSectionWrapper>
        </Stack>
    );
}

export default Editor;
