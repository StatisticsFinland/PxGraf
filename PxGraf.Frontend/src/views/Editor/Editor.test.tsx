import { render, screen } from "@testing-library/react";
import { ICubeMetaResult } from "api/services/cube-meta";
import { IHeaderResult } from "api/services/default-header";
import { IFilterVariableResult } from "api/services/filter-dimension";
import { IVisualizationSettingsResult } from "api/services/visualization-rules";
import { ISaveQueryResult } from "api/services/queries";
import React from "react";
import { VariableType } from "types/cubeMeta";
import Editor from "./Editor";
import { IQueryInfoResult } from "api/services/query-info";
import { HashRouter } from "react-router-dom";
import { IVisualizationResult } from "api/services/visualization";
import { EVariableType, ETimeVariableInterval, EVisualizationType } from "@statisticsfinland/pxvisualizer";
import { ITypeSpecificVisualizationRules } from "types/visualizationRules";
import serializer from "../../testUtils/stripHighchartsHashes";
import { NavigationProvider } from "contexts/navigationContext";
import { IValidateTableMetaDataResult } from "api/services/validate-table-metadata";
import { QueryClient, QueryClientProvider } from 'react-query';
import { UiLanguageContext } from "contexts/uiLanguageContext";

const queryClient = new QueryClient();

const language = "fi";
const setLanguage = jest.fn();
const languageTab = "fi";
const setLanguageTab = jest.fn();
const availableUiLanguages = ["fi", "sv", "en"];
const uiContentLanguage = "fi";
const setUiContentLanguage = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: () => {
        return {
            '*': 'foo/bar'
        };
    },
}));

jest.mock('react-i18next', () => ({
    ...jest.requireActual('react-i18next'),
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => ({})),
            },
        };
    },
}));

jest.mock('api/services/cube-meta', () => ({
    ...jest.requireActual('api/services/cube-meta'),
    useCubeMetaQuery: () => {
        return mockCubeMetaResult;
    }
}));

jest.mock('api/services/default-header', () => ({
    ...jest.requireActual('api/services/default-header'),
    useDefaultHeaderQuery: () => {
        return mockHeaderResult;
    }
}));

jest.mock('api/services/query-info', () => ({
    ...jest.requireActual('api/services/query-info'),
    useQueryInfoQuery: () => {
        return mockQueryInfoResult;
    }
}));

jest.mock('api/services/filter-dimension', () => ({
    ...jest.requireActual('api/services/filter-dimension'),
    useResolveVariableFiltersQuery: () => {
        return mockFilterVariableResult;
    }
}));

jest.mock('api/services/visualization-rules', () => ({
    ...jest.requireActual('api/services/visualization-rules'),
    useVisualizationOptionsQuery: () => {
        return mockVisualizationSettingsResult;
    }
}));

jest.mock('api/services/queries', () => ({
    ...jest.requireActual('api/services/queries'),
    useSaveMutation: () => {
        return mockSaveQueryResult;
    }
}));

jest.mock('api/services/visualization', () => ({
    ...jest.requireActual('api/services/visualization'),
    useVisualizationQuery: () => {
        return mockVisualizationResult;
    }
}));

jest.mock('@statisticsfinland/pxvisualizer', () => {
    const lib = jest.requireActual("@statisticsfinland/pxvisualizer");
    return {
        ...lib,
        Chart: (...args: any[]) => {
            return (
                <pre data-testid={'Chart'}>
                   args={JSON.stringify(args)}
                </pre>
            );
        }
    }
});

const mockVisualizationResult: IVisualizationResult = {
    isLoading: false,
    isError: false,
    data: {
        data: [ 1, 2, 3, 4 ],
        missingDataInfo: {},
        dataNotes: [],
        rowVariableCodes: ["foobar2"],
        columnVariableCodes: ["foobar1"],
        header: { fi: "foo", sv: "bar", en: "foobar" },
        selectableVariableCodes: ["foobar1"],
        visualizationSettings:
        {
            timeVariableIntervals: ETimeVariableInterval.Irregular,
            visualizationType: EVisualizationType.HorizontalBarChart,
            sorting: 'DESCENDING'
        },
        metaData: [{
            code: "foobar1",
            name: { fi: "foo1", sv: "bar1", en: "foobar1" },
            note: { fi: "föö1", sv: "bär1", en: "fööbär1" },
            type: EVariableType.Content,
            values: [{
                code: "barfoo1",
                name: { fi: "fyy1", sv: "bör1", en: "fyybör1" },
                note: { fi: "fuu1", sv: "baar1", en: "fuubaar1" },
                isSum: false,
                contentComponent: {
                    unit: { fi: "unitfi1", sv: "unitsv1", en: "uniten1" },
                    source: { fi: "sourcefi1", sv: "sourcesv1", en: "sourceen1" },
                    numberOfDecimals: 0,
                    lastUpdated: "2021-01-01T00:00:00Z"
                }
            }, {
                code: "barfoo2",
                name: { fi: "fyy2", sv: "bör2", en: "fyybör2" },
                note: { fi: "fuu2", sv: "baar2", en: "fuubaar2" },
                isSum: false,
                contentComponent: {
                    unit: { fi: "unitfi2", sv: "unitsv2", en: "uniten2" },
                    source: { fi: "sourcefi2", sv: "sourcesv2", en: "sourceen2" },
                    numberOfDecimals: 0,
                    lastUpdated: "2021-01-01T00:00:00Z"
                }
            }],
        }, {
            code: "foobar2",
            name: { fi: "foo2", sv: "bar2", en: "foobar2" },
            note: { fi: "föö2", sv: "bär2", en: "fööbär2" },
            type: EVariableType.OtherClassificatory,
            values: [{
                code: "barfoo1",
                name: { fi: "fyy1", sv: "bör1", en: "fyybör1" },
                note: { fi: "fuu1", sv: "baar1", en: "fuubaar1" },
                isSum: false,
                contentComponent: null,
            }, {
                code: "barfoo2",
                name: { fi: "fyy2", sv: "bör2", en: "fyybör2" },
                note: { fi: "fuu2", sv: "baar2", en: "fuubaar2" },
                isSum: false,
                contentComponent: null,
            }],
        }],
        tableReference: { hierarchy: ["foo", "bar"], name:  "foobar_table" },
    },
}

const mockSaveQueryResult: ISaveQueryResult = {
    isLoading: false,
    isError: false,
    isSuccess: true,
    data: {
        id: 'id'
    },
    mutate: jest.fn()
}

const mockTypeSpecificVisualizationRules: ITypeSpecificVisualizationRules = {
    allowShowingDataPoints: true,
    allowCuttingYAxis: true,
    allowMatchXLabelsToEnd: true,
    allowSetMarkerScale: true
}

const mockVisualizationSettingsResult: IVisualizationSettingsResult = {
    isLoading: false,
    isError: false,
    data: {
        allowManualPivot: false,
        multiselectVariableAllowed: false,
        sortingOptions: [
            {
                code: 'DESCENDING',
                description: {
                    'fi': 'Laskeva'
                }
            }
        ],
        visualizationTypeSpecificRules: mockTypeSpecificVisualizationRules
    }
}

const mockFilterVariableResult: IFilterVariableResult = {
    isLoading: false,
    isError: false,
    data: {
        'foo': ['bar', 'baz']
    }
}

const mockQueryInfoResult: IQueryInfoResult = {
    isLoading: false,
    isError: false,
    data: {
        maximumHeaderLength: 120,
        maximumSupportedSize: 100,
        size: 5,
        sizeWarningLimit: 75,
        validVisualizations: ['HorizontalBarChart', 'VerticalBarChart'],
        visualizationRejectionReasons: {
            'pieChart': {
                'fi': 'huono kaavio'
            }
        }
    }
}

const mockHeaderResult: IHeaderResult = {
    isLoading: false,
    isError: false,
    data: {
        'fi': 'header'
    }
}

const mockCubeMetaResult: ICubeMetaResult = {
    isLoading: false,
    isError: false,
    data: {
        header: {
            'fi': 'header'
        },
        languages: ['fi'],
        note: {
            'fi': 'note'
        },
        variables: [
            {
                code: 'code',
                name: {
                    'fi': 'variableName'
                },
                note: {
                    'fi': 'variableNote'
                },
                type: VariableType.Content,
                values: [
                    {
                        code: 'variableValueCode',
                        isSum: false,
                        name: {
                            'fi': 'variableValueName'
                        },
                        note: {
                            'fi': 'variableValueNote'
                        }
                    }
                ]
            }
        ]
    }
}

const mockTableValidationResult: IValidateTableMetaDataResult = {
    isLoading: false,
    isError: false,
    data: {
        tableHasContentVariable: true,
        tableHasTimeVariable: true,
        allVariablesContainValues: true
    }
}

const mockInvalidTableValidationResult: IValidateTableMetaDataResult = {
    isLoading: false,
    isError: false,
    data: {
        tableHasContentVariable: false,
        tableHasTimeVariable: false,
        allVariablesContainValues: false
    }
}

let mockResult: IValidateTableMetaDataResult;

jest.mock('api/services/validate-table-metadata', () => ({
    useValidateTableMetadataQuery: () => mockResult,
}));

describe('Rendering test', () => {
    beforeAll(() => {
        expect.addSnapshotSerializer(serializer);
        mockResult = mockTableValidationResult;
    });

    it('renders correctly', () => {
        const { asFragment } = render(
            <QueryClientProvider client={queryClient}>
                <NavigationProvider>
                    <HashRouter>
                        <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                            <Editor />
                        </UiLanguageContext.Provider>
                    </HashRouter>
                </NavigationProvider>
            </QueryClientProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('Assertion tests', () => {
    it('renders errorContainer with correct message when tableValidityResponse is invalid', () => {
        mockResult = mockInvalidTableValidationResult;
        render(
            <QueryClientProvider client={queryClient}>
                <NavigationProvider>
                    <HashRouter>
                        <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                            <Editor />
                        </UiLanguageContext.Provider>
                    </HashRouter>
                </NavigationProvider>
            </QueryClientProvider>
        );

        expect(screen.getByText('error.contentVariableMissing error.timeVariableMissing error.variablesMissingValues')).toBeInTheDocument();
    });
});