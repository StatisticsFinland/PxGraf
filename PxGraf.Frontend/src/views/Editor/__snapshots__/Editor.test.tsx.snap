// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering test renders correctly 1`] = `
<DocumentFragment>
  <div
  class="MuiStack-root css-m69qwo-MuiStack-root"
>
  <div
    class="sc-eJgwjL gwNfXU MuiBox-root css-ve9ogo"
  >
    <div
      class="sc-jzjKHc hOhaiK"
    >
      <p
        class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
      >
        variableSelect.title
      </p>
      <button
        aria-label="tooltip.open: variableSelect.title"
        class="sc-dkjaqt fTFHez"
        id="mainContent"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorInfo MuiSvgIcon-fontSizeMedium css-1yq1soz-MuiSvgIcon-root"
          data-testid="InfoIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 15h-2v-6h2zm0-8h-2V7h2z"
          />
        </svg>
      </button>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded Mui-expanded MuiAccordion-gutters sc-tOkKi diPSiL css-1086bdv-MuiPaper-root-MuiAccordion-root"
    >
      <div
        aria-controls="panel1a-content"
        aria-expanded="true"
        class="MuiButtonBase-root MuiAccordionSummary-root Mui-expanded MuiAccordionSummary-gutters css-1f773le-MuiButtonBase-root-MuiAccordionSummary-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiAccordionSummary-content Mui-expanded MuiAccordionSummary-contentGutters css-eqpfi5-MuiAccordionSummary-content"
        >
          <h2
            class="MuiTypography-root MuiTypography-h2 css-1sra7t5-MuiTypography-root"
          >
            <b>
              variableName 0/1
            </b>
          </h2>
        </div>
        <div
          class="MuiAccordionSummary-expandIconWrapper Mui-expanded css-yw020d-MuiAccordionSummary-expandIconWrapper"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ExpandMoreIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-entered css-pwcg7p-MuiCollapse-root"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
          >
            <div
              class="MuiAccordion-region"
              id="panel1a-content"
              role="region"
            >
              <div
                class="MuiAccordionDetails-root sc-czgmHJ kLUNZd css-15v22id-MuiAccordionDetails-root"
              >
                <div
                  class="MuiStack-root sc-kkmypM kJaLMI css-1sazv7p-MuiStack-root"
                >
                  <div
                    class="MuiStack-root sc-dnaUSb dlJuVl css-m69qwo-MuiStack-root"
                  >
                    <span
                      aria-label="variableSelect.openMenu"
                      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                      role="button"
                      style="padding-left: 0px; margin-right: 12px;"
                      tabindex="0"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="MenuIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M3 18h18v-2H3zm0-5h18v-2H3zm0-7v2h18V6z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </span>
                    <div
                      class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon sc-iKTcqh jsvyIL css-1x6bjyf-MuiAutocomplete-root"
                    >
                      <div
                        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                      >
                        <label
                          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                          data-shrink="false"
                          for=":r0:"
                          id=":r0:-label"
                        >
                          variableSelect.itemFilter
                        </label>
                        <div
                          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
                        >
                          <input
                            aria-autocomplete="list"
                            aria-expanded="false"
                            aria-invalid="false"
                            autocapitalize="none"
                            autocomplete="off"
                            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                            id=":r0:"
                            role="combobox"
                            spellcheck="false"
                            type="text"
                            value=""
                          />
                          <div
                            class="MuiAutocomplete-endAdornment css-p1olib-MuiAutocomplete-endAdornment"
                          >
                            <button
                              aria-label="selectable.open"
                              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                              tabindex="-1"
                              title="selectable.open"
                              type="button"
                            >
                              <svg
                                aria-hidden="t...
</DocumentFragment>
`;
